Prevent systems using nvme-remap from crossing checkpoints

wjt opened the pull request:
- Remove all previous checkpoint checks
- Factor out the should_follow_checkpoint() function so it can be tested in isolation
- Add a new check that prevents upgrades if nvme-remap driver is in use

~~I don't actually have this hardware so this is only tested by the test case; and it's a Friday evening; hence draft.~~

I targeted `eos5.1` with this PR because that's where the code needs to land to block upgrading to `eos6.0`; we'd merge to master after landing on `eos5.1`, so the same commits (not cherry-picks) will be on both.

https://phabricator.endlessm.com/T35247
https://phabricator.endlessm.com/T35250

wjt commented on libeos-updater-util/checkpoint.c:
```diff
@@ -0,0 +1,145 @@
+}
+
+static gboolean
+is_nvme_remap_in_use (OstreeSysroot *sysroot)
```
I don't actually have the relevant hardware. I will build some packages and ask those in Taipei to check them, if this doesn't land before that.

wjt replied:
@starnight confirmed that this works correctly on the device he has that uses `nvme-remap`. Thanks!

dbnicholson commented on libeos-updater-util/tests/checkpoint.c:
```diff
@@ -60,6 +59,7 @@ teardown (Fixture       *fixture,
 {
   g_clear_object (&fixture->sysroot);
   g_clear_object (&fixture->repo);
+  g_clear_object (&fixture->root_dir);
```
I think this and the `g_unsetenv` below should be rolled into the previous commit.

dbnicholson commented on libeos-updater-util/tests/checkpoint.c:
```diff
@@ -89,6 +89,69 @@ test_force_no_follow (Fixture                  *fixture,
   g_assert_false (euu_should_follow_checkpoint (fixture->sysroot, "os/eos/amd64/latest2", "os/eos/amd64/latest3", &reason));
   g_assert_cmpstr (reason, !=, NULL);
+
+  g_unsetenv ("EOS_UPDATER_FORCE_FOLLOW_CHECKPOINT");
```
Probably wouldn't be a bad idea to run this also in `setup()` to ensure it's not inherited from the environment for tests that don't set it.

dbnicholson commented on eos-updater/poll-common.c:
```diff
@@ -407,20 +380,77 @@ cleanstr (gchar *s)
+static gboolean
+stop_aggregate_timer (gpointer data)
+{
+  g_return_val_if_fail (EMTR_IS_AGGREGATE_TIMER (data), G_SOURCE_REMOVE);
```
Maybe the compiler will elide this function, but I'd assume this function needs to be under `
#ifdef HAS_EOSMETRICS_0` for the `emtr` symbols and macros.

wjt replied:
Good point!

The other option is to not wait 1 second and just stop it immediately after it starts. The aggregate will still get reported even if its count is 0…

wjt replied:
I moved the `#ifdef` out to encompass both functions, and added an `#else` branch that defines a stub `maybe_send_checkpoint_blocked_event()`.

dbnicholson commented on eos-updater/poll-common.c:
```diff
@@ -407,20 +380,77 @@ cleanstr (gchar *s)
+    {
+      g_message ("Skipping metric event %s: %s (metrics disabled)",
+               EOS_UPDATER_METRIC_CHECKPOINT_BLOCKED,
+               payload_str);
```
I presume this was `g_debug` before and now the alignment is off.

wjt replied:
Yes, I changed this for debugging purposes and didn't mean to commit it.

dbnicholson reviewed (changes requested):
Looks good. I left some minor issues, but overall I like it. I think the pseudo test in the last commit leaves a bit to be desired:

* We don't actually install the `libeos-updater-util-0-tests` package. It would be nice to be able to do the check without having to alter the system.
* The refs are hardcoded, so it would make it hard to test if we had checkpoints that only needed to be crossed on certain upgrade paths.

That said, it's good enough.

wjt commented on libeos-updater-util/checkpoint.c:
```diff
@@ -0,0 +1,70 @@
+  g_return_val_if_fail (booted_ref != NULL, FALSE);
+  g_return_val_if_fail (target_ref != NULL, FALSE);
+  g_return_val_if_fail (out_reason != NULL, FALSE);
+  g_return_val_if_fail (*out_reason == NULL, FALSE);
```
Good point. This does get deleted in a subsequent patch but I will amend it.

wjt replied:
Actually my original code is correct. `out_reason` has type `char **`. The first assertion tests that the caller has supplied a non-NULL pointer as the location to store a pointer to the reason. The second assertion is that that location currently holds a NULL pointer, so that assigning to `*out_reason` will not leak whatever string was previously there.

dbnicholson replied:
Ah, correct on both accounts.

wjt commented:
The pseudo-test was done like that for convenience. I guess you're right and I should do the small amount of buildsystem work to install `/usr/libexec/eos-updater-check-checkpoint-checks` or similar. Honestly one reason I didn't is that I can't decide what to call it.

wjt commented:
I moved the dodgy fake test to a standalone `eos-updater-test-checkpoint` executable, installed as part of the `eos-updater` package. I am not that keen on the name – it doesn't test the checkpoint, it tests the checks that are applied at the checkpoint – but, good enough? For good measure I also added a manpage, and hope to never write groff source again in my lifetime.

dbnicholson commented on eos-updater-test-checkpoint/meson.build:
```diff
@@ -0,0 +1,24 @@
+  install_dir: get_option('libexecdir'),
+)
+
+install_man('docs/eos-updater-test-checkpoint.8')
```
Missing newline here.

wjt replied:
This bad default is one of the reasons I added an `.editorconfig` in #335 :)

dbnicholson commented on eos-updater-test-checkpoint/meson.build:
```diff
@@ -0,0 +1,24 @@
+]
+
+eos_updater_test_checkpoint_deps = [
+  dependency('ostree-1', version: '>= 2019.2'),
```
Can you change this to `ostree_dep` per 0104703e68bcb2442b386aa43beb27b0ec1ee6c2?

wjt replied:
I can't, because that patch isn't on `eos5.1`. Oh, but it's the very next commit after the current `eos5.1` so I'll rebase this branch onto it.

dbnicholson replied:
Ah ha. I forgot about the eos5.1 base.

dbnicholson reviewed (approved):
Looks great. Way to roff, by the way. I can't think of a better name for the program, so this all looks fine to me modulo the last couple nitpicks.

wjt commented on eos-updater/poll-common.c:
```diff
@@ -406,21 +380,83 @@ cleanstr (gchar *s)
 {
   g_autoptr(GHashTable) hw_descriptors = NULL;
-  g_autoptr(EosMetricsInfo) info = NULL;
+  g_autofree gchar *vendor, *product;
```
I think they're assigned to on all paths so it's safe not to, but you're right, it would be clearer to do thay
